File: ts/kitstart/src/core/seo/sitemap.ts
import { isPublished } from "../place/publication";
import type { Place } from "../place/types";
import { placeOrigin, siteOrigin } from "../place/view";
import { createRouting } from "../routing";
import type { Site } from "../site";

/**
 * One sitemap per host: a sitemap may only list URLs on its own host, and a
 * place's canonical host is its subdomain (the apex on a `single` site). The
 * apex of a `subdomains` site lists the brand page.
 *
 * Each language version is its own entry, annotated with the whole set
 * *including itself*. No `lastmod`: a deploy timestamp that moves without the
 * content moving is a lie. An unpublished place is left out entirely, and a
 * site with no domain lists nothing.
 *
 * Shapes match Next's `MetadataRoute.Sitemap` / `MetadataRoute.Robots`, so
 * `./next` returns them as they are.
 */
export interface SitemapEntry {
  url: string;
  priority: number;
  alternates: { languages: Record<string, string> };
}

export interface RobotsRule {
  userAgent: string | string[];
  allow?: string;
  disallow?: string;
}

export interface RobotsFile {
  rules: RobotsRule[];
  sitemap?: string;
}

function cluster<L extends string, P extends string>(site: Site<L, P>, origin: string, suffix: string, priority: number): SitemapEntry[] {
  const languages = site.i18n.languageAlternates(suffix || "/", origin);
  return site.i18n.locales.map(locale => ({
    url: `${origin}${site.i18n.localePath(locale, suffix || "/")}`,
    priority,
    alternates: { languages },
  }));
}

/**
 * The sitemap for a request's host. `places` is the list as the live source
 * sees it now — the caller fetches it strictly: an unreachable source must
 * fail the sitemap (a 5xx a crawler retries), never shrink it (a list of
 * places a crawler drops).
 */
export function sitemapFor<L extends string, P extends string>(site: Site<L, P>, host: string, places: readonly Place<L>[]): SitemapEntry[] {
  const apex = siteOrigin(site);
  if (apex === null) return [];
  const slug = createRouting(site).hostSlug(host);
  if (slug === null) return cluster(site, apex, "", 1);
  const place = places.find(p => p.slug === slug);
  const origin = placeOrigin(site, slug);
  if (!place || origin === null || !isPublished(place, site.publication, site.brand)) return [];
  return site.pageKeys.flatMap(page => cluster(site, origin, site.pages[page], page === "home" ? 1 : 0.8));
}

/** Crawlers named explicitly: several treat a bare wildcard as ambiguous. */
export const AI_CRAWLERS = ["GPTBot", "ClaudeBot", "PerplexityBot", "Google-Extended"];

/**
 * Allow everything, and point at the host's own sitemap. Unpublished places
 * are held back by `noindex`, not here, so a crawler can still read the
 * `noindex`. A site with no domain is not ready to be crawled at all.
 */
export function robotsFor<L extends string, P extends string>(site: Site<L, P>, host: string): RobotsFile {
  const apex = siteOrigin(site);
  if (apex === null) return { rules: [{ userAgent: "*", disallow: "/" }] };
  const slug = createRouting(site).hostSlug(host);
  const origin = (slug === null ? null : placeOrigin(site, slug)) ?? apex;
  return {
    rules: [
      { userAgent: "*", allow: "/" },
      { userAgent: [...AI_CRAWLERS], allow: "/" },
    ],
    sitemap: `${origin}/sitemap.xml`,
  };
}

/** The OG card URL: on the apex, so it never needs a rewrite. */
export function ogImageUrl<L extends string, P extends string>(site: Site<L, P>, query: { slug?: string; locale: L; page?: string }): string {
  const params = new URLSearchParams({ lang: query.locale });
  if (query.slug) params.set("l", query.slug);
  if (query.page) params.set("p", query.page);
  return `${siteOrigin(site) ?? ""}/og?${params.toString()}`;
}

File: rust/src/kitstart/sitemap.rs
//! One sitemap per host: a sitemap may only list URLs on its own host, and a
//! place's canonical host is its subdomain (the apex on a `Single` site). The
//! apex of a `Subdomains` site lists the brand page.
//!
//! Each locale version is its own entry, annotated with the whole cluster
//! *including itself*. No `lastmod`: a deploy timestamp that moves without the
//! content moving is a lie. An unpublished place is left out entirely, and a
//! site with no domain lists nothing.

use super::{
	json::{Json, Object},
	place::Place,
	routing::host_slug,
	site::Site,
};

#[derive(Clone, Debug, PartialEq)]
pub struct SitemapEntry {
	pub url: String,
	pub priority: f64,
	/// `(hreflang, url)`, `x-default` last.
	pub languages: Vec<(String, String)>,
}

/// Crawlers named explicitly: several treat a bare wildcard as ambiguous.
pub const AI_CRAWLERS: [&str; 4] = ["GPTBot", "ClaudeBot", "PerplexityBot", "Google-Extended"];

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum RobotsFile {
	/// Allow everyone, the named AI crawlers explicitly, and point at `sitemap`.
	Open { sitemap: String },
	/// A site with no domain is not ready to be crawled at all.
	Closed,
}

fn cluster(site: &Site, origin: &str, suffix: &str, priority: f64) -> Vec<SitemapEntry> {
	let path = if suffix.is_empty() { "/" } else { suffix };
	let languages = site.i18n().language_alternates(path, origin);
	site.i18n()
		.locales()
		.map(|locale| SitemapEntry {
			url: format!("{origin}{}", site.i18n().locale_path(locale, path)),
			priority,
			languages: languages.clone(),
		})
		.collect()
}

/// The sitemap for a request's host. `places` is the list as the live source
/// sees it now — fetched strictly by the caller: an unreachable source must
/// fail the sitemap (a 5xx a crawler retries), never shrink it (a list of
/// places a crawler drops).
pub fn sitemap_for(site: &Site, host: &str, places: &[Place]) -> Vec<SitemapEntry> {
	let Some(apex) = site.origin() else {
		return Vec::new();
	};
	let Some(slug) = host_slug(site, host) else {
		return cluster(site, &apex, "", 1.0);
	};
	let Some(place) = places.iter().find(|p| p.slug == slug) else {
		return Vec::new();
	};
	let Some(origin) = site.place_origin(slug) else {
		return Vec::new();
	};
	if !site.publication().is_published(place, site.brand().domain.as_deref()) {
		return Vec::new();
	}
	site.pages()
		.flat_map(|page| cluster(site, &origin, page.suffix, if page.is_home() { 1.0 } else { 0.8 }))
		.collect()
}

/// Allow everything, and point at the host's own sitemap. Unpublished places
/// are held back by `noindex`, not here, so a crawler can still read the
/// `noindex`.
pub fn robots_for(site: &Site, host: &str) -> RobotsFile {
	let Some(apex) = site.origin() else {
		return RobotsFile::Closed;
	};
	let origin = host_slug(site, host).and_then(|slug| site.place_origin(slug)).unwrap_or(apex);
	RobotsFile::Open {
		sitemap: format!("{origin}/sitemap.xml"),
	}
}

/// The Next `MetadataRoute.Sitemap` shape.
pub fn sitemap_json(entries: &[SitemapEntry]) -> Json {
	entries
		.iter()
		.map(|entry| {
			let languages: Object = entry.languages.iter().map(|(tag, url)| (tag.clone(), Json::from(url))).collect();
			Json::from(
				Object::new()
					.with("url", &entry.url)
					.with("priority", entry.priority)
					.with("alternates", Object::new().with("languages", languages)),
			)
		})
		.collect::<Vec<_>>()
		.into()
}

/// `sitemap.xml`, with every entry's cluster as `xhtml:link` alternates.
pub fn sitemap_xml(entries: &[SitemapEntry]) -> String {
	let mut out = String::from("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<urlset xmlns=\"http://www.sitemaps.org/schemas/sitemap/0.9\" xmlns:xhtml=\"http://www.w3.org/1999/xhtml\">\n");
	for entry in entries {
		out.push_str(&format!("  <url>\n    <loc>{}</loc>\n", xml_escape(&entry.url)));
		for (tag, url) in &entry.languages {
			out.push_str(&format!("    <xhtml:link rel=\"alternate\" hreflang=\"{}\" href=\"{}\"/>\n", xml_escape(tag), xml_escape(url)));
		}
		out.push_str(&format!("    <priority>{:.1}</priority>\n  </url>\n", entry.priority));
	}
	out.push_str("</urlset>\n");
	out
}

fn xml_escape(s: &str) -> String {
	s.replace('&', "&amp;").replace('<', "&lt;").replace('>', "&gt;").replace('"', "&quot;").replace('\'', "&apos;")
}

impl RobotsFile {
	/// The Next `MetadataRoute.Robots` shape.
	pub fn to_json(&self) -> Json {
		match self {
			Self::Open { sitemap } => Object::new()
				.with(
					"rules",
					vec![
						Object::new().with("userAgent", "*").with("allow", "/"),
						Object::new().with("userAgent", AI_CRAWLERS.iter().map(|c| Json::from(*c)).collect::<Vec<_>>()).with("allow", "/"),
					],
				)
				.with("sitemap", sitemap)
				.into(),
			Self::Closed => Object::new().with("rules", vec![Object::new().with("userAgent", "*").with("disallow", "/")]).into(),
		}
	}

	/// `robots.txt`.
	pub fn to_txt(&self) -> String {
		match self {
			Self::Open { sitemap } => {
				let named: String = AI_CRAWLERS.iter().map(|c| format!("User-agent: {c}\n")).collect();
				format!("User-agent: *\nAllow: /\n\n{named}Allow: /\n\nSitemap: {sitemap}\n")
			}
			Self::Closed => "User-agent: *\nDisallow: /\n".to_owned(),
		}
	}
}
